extra_data/reader.py

```diff
@@ -289,18 +292,39 @@ def __init__(self, files, selection=None, train_ids=None, ctx_closes=False):
             train_ids = sorted(set().union(*(f.train_ids for f in files)))
         self.train_ids = train_ids
 
+    @staticmethod
+    def _open_file(path, cache_info=None):
+        try:
+            fa = FileAccess(path, _cache_info=cache_info)
+        except Exception as e:
+            print("Skipping file", path, file=sys.stderr)
+            print("  (error was: {})".format(e), file=sys.stderr)
+        else:
+            return fa
+
     @classmethod
     def from_paths(cls, paths, _files_map=None):
         files = []
+        uncached = []
         for path in paths:
             cache_info = _files_map and _files_map.get(path)
-            try:
-                fa = FileAccess(path, _cache_info=cache_info)
-            except Exception as e:
-                print("Skipping file", path, file=sys.stderr)
-                print("  (error was: {})".format(e), file=sys.stderr)
+            if cache_info:
+                fa = cls._open_file(path, cache_info=cache_info)
+                if fa is not None:
+                    files.append(fa)
             else:
-                files.append(fa)
+                uncached.append(path)
+
+        if uncached:
+            def initializer():
+                # prevent child processes from receiving KeyboardInterrupt
+                signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+            nproc = min(len(psutil.Process().cpu_affinity()), len(uncached))
```

Comment:
There a different scenarios:
- data is on gpfs and not cached on the node: more processes is better (usual case where multiprocessing is used), see few tests:

| # files | # process | time (s) |
|---------|-----------|----------|
| 207     | 10        | 7.49     |
| 207     | 20        | 4.42     |
| 207     | 30        | 3.57     |
| 207     | 40        | 2.69     |
| 207     | 80        | 2.27     |

- data is on gpfs and cached locally: not much influence. I support that in this case, the `map` files are already generated so multiprocessing won't be used.

- data on `archive`: I don't know if that's just today, but it is insanely slow! maybe I should test again later (do you have any experience?) - we absolutely need to have the map files here (I remember that they are invalidated once data is moved to archive :-( )

Reply:
OK, so it looks like we can continue to benefit from more processes, even if not linearly. Let's leave this as is, then, and see if anything comes up that suggests a cap would be sensible. Thanks for investigating this systematically.

Archived data: as I understand it, once the data is archived on dCache, it may need to be read from tape, which may be arbitrarily slow, especially if the tape machines area already in use. I don't know if there's a convenient way to check 'is this file cached on disk?' without attempting to read it.

At the moment, even listing a directory seems to be ridiculously slow, though. Last time that happened, I think it was something Robert did...

Reply:
>  Last time that happened, I think it was something Robert did...

haha :D
OK so it is not a normal situation, I'll check tomorrow and email IT if that does not improve

Reply:
@RobertRosca what did you do last time? :o

Reply:
Nitpick: let's have a comment to clarify that cpu_affinity gives us cores we can use, which may be a subset of the cores the machine has. I don't think that's obvious without reading psutil's docs.